client/my-sites/checkout/cart/cart-coupon.jsx

```diff
@@ -37,87 +34,125 @@ class CartCoupon extends React.Component {
+	}
 
-		this.setState( {
-			userChangedCoupon: false,
-			hasSubmittedCoupon: true,
-		} );
-		applyCoupon( this.state.couponInputValue );
-	};
+	get appliedCouponCode() {
+		return this.props.cart.is_coupon_applied && this.props.cart.coupon
+			? this.props.cart.coupon
+			: null;
+	}
 
-	handleCouponInput = event => {
-		this.setState( {
-			userChangedCoupon: true,
-			couponInputValue: event.target.value,
-		} );
-	};
+	renderAppliedCoupon() {
+		return (
+			<div className="cart__coupon">
+				<span className="cart__details">
+					{ this.props.translate( 'Coupon applied: %(coupon)s', {
+						args: { coupon: this.appliedCouponCode },
+					} ) }
+				</span>{' '}
+				<button onClick={ this.clearCoupon } className="button is-link cart__remove-link">
+					{ this.props.translate( 'Remove' ) }
+				</button>
+			</div>
+		);
+	}
 
-	getToggleLink = () => {
+	renderApplyCouponUI() {
 		if ( this.props.cart.total_cost === 0 ) {
 			return;
 		}
 
-		if ( this.state.hasSubmittedCoupon ) {
-			return;
-		}
-
 		return (
-			<a href="" onClick={ this.toggleCouponDetails }>
-				{ this.props.translate( 'Have a coupon code?' ) }
-			</a>
+			<div className="cart__coupon">
+				<button onClick={ this.toggleCouponDetails } className="button is-link cart__toggle-link">
+					{ this.props.translate( 'Have a coupon code?' ) }
+				</button>
+
+				{ this.renderCouponForm() }
+			</div>
 		);
-	};
+	}
 
-	getCouponForm = () => {
+	renderCouponForm = () => {
 		if ( ! this.state.isCouponFormShowing ) {
 			return;
 		}
 
-		if ( this.state.hasSubmittedCoupon ) {
-			return;
-		}
-
 		return (
-			<form onSubmit={ this.applyCoupon }>
+			<form onSubmit={ this.applyCoupon } className={ 'cart__form' }>
 				<input
 					type="text"
+					disabled={ this.isSubmitting }
 					placeholder={ this.props.translate( 'Enter Coupon Code', { textOnly: true } ) }
-					onChange={ this.handleCouponInput }
+					onChange={ this.handleCouponInputChange }
 					value={ this.state.couponInputValue }
+					autoFocus // eslint-disable-line jsx-a11y/no-autofocus
 				/>
-				<Button type="submit" disabled={ isEmpty( trim( this.state.couponInputValue ) ) }>
+				<Button
+					type="submit"
+					disabled={ isEmpty( trim( this.state.couponInputValue ) ) }
+					busy={ this.isSubmitting }
+				>
 					{ this.props.translate( 'Apply' ) }
 				</Button>
 			</form>
 		);
 	};
 
-	render() {
-		return (
-			<div className="cart-coupon">
-				{ this.getToggleLink() }
-				{ this.getCouponForm() }
-			</div>
-		);
+	get isSubmitting() {
+		return ! this.props.cart.is_coupon_applied && this.props.cart.coupon ? true : false;
 	}
+
+	toggleCouponDetails = event => {
+		event.preventDefault();
+
+		this.setState( { isCouponFormShowing: ! this.state.isCouponFormShowing } );
+
+		if ( this.state.isCouponFormShowing ) {
+			analytics.ga.recordEvent( 'Upgrades', 'Clicked Hide Coupon Code Link' );
+		} else {
+			analytics.ga.recordEvent( 'Upgrades', 'Clicked Show Coupon Code Link' );
+		}
```

Comment:
This only impacts analytics, but looks like a risky use of setState. I'd suggest:

```js
		this.setState( ( { isCouponFormShowing } ) => ( { isCouponFormShowing: ! isCouponFormShowing } ), () => {
			if ( this.state.isCouponFormShowing ) {
				analytics.ga.recordEvent( 'Upgrades', 'Clicked Hide Coupon Code Link' );
			} else {
				analytics.ga.recordEvent( 'Upgrades', 'Clicked Show Coupon Code Link' );
			}			
		} );
```

> setState() does not always immediately update the component. It may batch or defer the update until later. This makes reading this.state right after calling setState() a potential pitfall.

https://reactjs.org/docs/react-component.html#setstate

Reply:
I reordered some methods in this component - I just moved these lines from above and didn't actually add them. It could be a good idea to refactor this component some more, but it's out of scope of this PR